nixos/nginx: default kTLS to on, disable sendfile when http2 is used to avoid performance regressions

SuperSandro2000 opened the pull request:
## Description of changes

As written in the description, kTLS improved performance significantly.

All supported nginx versions are older than 1.21.4 and all supported kernels are at least 4.13,
so turning this on by default should be safe.

see e.g. https://tinselcity.github.io/SSL_Sendfile/ for the version numbers

I have this option turned on for a long time without any issues to my knowledge.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 committed:
nixos/nginx: default kTLS to on

As written in the description, kTLS improved performance significantly.

All supported nginx versions are older than 1.21.4 and all supported kernels are at least 4.13,
so turning this on by default should be safe.

see e.g. https://tinselcity.github.io/SSL_Sendfile/ for the version numbers

SuperSandro2000 committed:
nixos/nginx: drop description parts regarding old nginx versions

Shawn8901 reviewed (approved):
having that option also enabled on all my nginx instances, so i am fine with changing the default here.

SuperSandro2000 commented:
@RaitoBezarius can we merge this? I am not aware of any drawbacks.

RaitoBezarius commented:
> @RaitoBezarius can we merge this? I am not aware of any drawbacks.

What's your testing method?

Shawn8901 commented:
@RaitoBezarius do you mean how to verify in general or the benefits?

Testing in general:
To be sure that kTLS was picked up one could check debug logs for nginx. Sadly we dont have debug enabled by default, so we need to override that with `.override { withDebug = true; }`. 
We can then set `services.nginx.logError = "stderr debug";` so that logs for debug logging show up in journalctl.

https://docs.nginx.com/nginx/admin-guide/monitoring/debugging/#writing-the-debugging-log-to-a-file

Also we need a vhost that has  `kTLS = true;` set. 

Then the following logs should show up.


```log
Jun 11 21:36:29 tank nginx[11169]: 2024/06/11 21:36:29 [debug] 11169#11169: *1 BIO_get_ktls_send(): 1
```

for performance things possibly that could help 
https://www.f5.com/company/blog/nginx/improving-nginx-performance-with-kernel-tls
i did not do any performance analysis by myself, but the article claims somewhat about ~13% ob ubuntu 21.10 in throughput when `sendfile` is used.


I hope that answers the question and helps.

Izorkin commented:
Note, that when using sendfile, the download speed of files via the HTTP/2 protocol decreases.

FliegendeWurst reviewed (commented):
> Note, that when using sendfile, the download speed of files via the HTTP/2 protocol decreases.

I found a relevant [ML thread](https://mailman.nginx.org/pipermail/nginx-devel/2021-December/014600.html): ([1](https://mailman.nginx.org/pipermail/nginx-devel/2021-December/014601.html))
> That's expected, because of HTTP/2 framing.  Unfortunately, HTTP/2 
isn't designed to work with sendfile(), and sending large files 
over HTTP/2 require a lot of sendfile() syscalls.  In general, for 
HTTP/2 it is better to keep sendfile() disabled.

SuperSandro2000 commented:
I am going to create a 2nd PR for that.

SuperSandro2000 committed:
nixos/nginx: move sendfile to virtualHost to allow it easily be disabled when http2 is enabled

Otherwise we would run into regressions when enabling kTLS
https://mailman.nginx.org/pipermail/nginx-devel/2021-December/014601.html
https://mailman.nginx.org/pipermail/nginx-devel/2021-December/014603.html

Izorkin commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -373,11 +372,16 @@ let
           server {
             ${concatMapStringsSep "\n" listenString redirectListen}
-
+            ${sendfile}
```
This will only work for the block `server` with port 80. You need to add it to another block as well.

Izorkin commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -373,11 +372,16 @@ let
+        # sendfile works well with kTLS/SSL_sendfile() from openssl but not with http2
+        # https://mailman.nginx.org/pipermail/nginx-devel/2021-December/014601.html
+        # https://mailman.nginx.org/pipermail/nginx-devel/2021-December/014603.html
+        sendfile = lib.optionalString (cfg.recommendedOptimisation && !(hasSSL && vhost.http2)) "sendfile on;";
```
Maybe rename it to `enableSendfile`? To avoid confusion with the options that nginx uses.
